File: src/c++/network_flow/bipartite_matching/1867/main.cc
/* Date : 2019.5.10
 * Author : seungyeon choi
 * 1867 (돌멩이 제거)
 */

#include <cstdio>
#include <vector>

using namespace std;

const int MAX_COUNT = 501;

int N, K;
int adj[MAX_COUNT][MAX_COUNT];
vector<int> Achk, Bchk;
vector<bool> visit;

bool dfs(int i) {
  if (visit[i]) return false;
  visit[i] = true;
  for (int j = 0; j < N; j++) {
    if (adj[i][j]) {
      if (Bchk[j] == -1 || dfs(Bchk[j])) {
        Achk[i] = j;
        Bchk[j] = i;
        return true;
      }
    }
  }
  return false;
}

int bmatch() {
  Achk = vector<int>(N, -1);
  Bchk = vector<int>(N, -1);
  int ans = 0;
  for (int i = 0; i < N; i++) {
    visit = vector<bool>(N, false);
    if (dfs(i))
      ans++;
  }
  return ans;
}

int main() {
  int num1, num2;
  scanf("%d %d", &N, &K);
  for (int i = 0; i < K; i++) {
    scanf("%d %d", &num1, &num2);
    adj[num1 - 1][num2 - 1] = 1;
  }
  printf("%d\n", bmatch());
  return 0;
}

File: src/python/network_flow/bipartite_matching/1867/main.py
# Date : 2017.11.24
# Author : seungyeon choi
# 1867 (돌멩이 제거)

ans = 0
N, K = map(int, input().split())
graph = [[] for _ in range(N)]
Achk = [-1 for _ in range(N)]
Bchk = [-1 for _ in range(N)]
for _ in range(K):
    A, B = map(int, input().split())
    graph[A-1].append(B-1)
def dfs(i):
    global visit
    if visit[i] : return False
    visit[i] = 1
    for j in range(len(graph[i])):
        temp = graph[i][j]
        if (Bchk[temp] == -1) or dfs(Bchk[temp]) :
            Achk[i] = temp
            Bchk[temp] = i
            return True
    return False
for j in range(N):
    visit = [0 for _ in range(N)]
    if dfs(j):
        ans += 1
print (ans)
